fix(theme): make Prism code block language / additionalLanguages case insensitive

heysujal opened the pull request:
## Pre-flight checklist

- [X] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.

## Motivation
fix #9012 

If someone uses a language that is not supported by Docusaurus by [default](https://docusaurus.io/docs/markdown-features/code-blocks#supported-languages) then you have to add that language in `additionalLanguages[]`  in `docusaurus.config.js`. If a language added is not in lowercase, then highlighting for that language will not work. For example
 Adding `'PHP'` or `'pHp'` instead of `'php'`

## Test Plan
0. Good : `additionalLanguages: ['php'],`

![image](https://github.com/facebook/docusaurus/assets/55016909/d9697d8b-dc1c-4d61-ada7-0d91fd4343d1)


1. Bad : `additionalLanguages: ['PHP'],`

![image](https://github.com/facebook/docusaurus/assets/55016909/625ff6d0-f227-4f9b-9b35-112a2d65b3e8)

2. Bad : `additionalLanguages: ['pHp'],`

![image](https://github.com/facebook/docusaurus/assets/55016909/7a49eeee-0339-482b-ba9c-6a9effa04575)



### Test links

<!--
🙏 Please add an exhaustive list of links relevant to this pull request.
⏱ This saves maintainers a lot of time during reviews.

- If you changed anything that's displayed on UI, please add a dogfooding page in website/_dogfooding to help us preview the effect. Those tests are deployed at https://docusaurus.io/tests
- If you changed documentation, please link to the new and updated documentation pages.

After submission, our Netlify bot will post a deploy preview link in comment, in the format of https://deploy-preview-<PR-NUMBER>--docusaurus-2.netlify.app/. Once available, please edit this section with links to the relevant deploy preview pages.

Please don't be afraid to change the main site's configuration as well! You can make use of your new feature on our site so we can preview its effects. We can decide if it should be kept in production before merging it.
-->

https://deploy-preview-9183--docusaurus-2.netlify.app/tests/pages/code-block-tests/#code-block-prism-language-tests

## Related issues/PRs
https://github.com/facebook/docusaurus/issues/9012

heysujal committed:
add toLowerCase()

Signed-off-by: Sujal Gupta <sujalgupta6100@gmail.com>

heysujal committed:
add test code block for good/bad prism language case

Signed-off-by: Sujal Gupta <sujalgupta6100@gmail.com>

heysujal committed:
Merge branch 'facebook:main' into case-agnostic-highlight

heysujal committed:
add PR link

slorber commented on website/_dogfooding/_pages tests/code-block-tests.mdx:
````diff
@@ -27,6 +27,21 @@ See:
 
+## Code block with/without the good prism language case in additionalLanguages[] tests
+
+```php
````
that would be nice to add 2 other cases like `PHP` and `PhP` there to ensure we also support case insensitive language metastring

heysujal replied:
Well, i already tried doing that before making a pr.
But only php code block works and other like you mentioned doesn't highlight code

heysujal replied:
Well, I have added them and I also saw that the VSCode Markdown preview supports highlighting all these blocks. So, maybe we should do it too.

slorber reviewed (changes requested):
Thanks

You can keep `phP` or something like that in the additional languages of this PR so that we can make PHP work in the preview and be able to validate the PR without running local modifications. We are tracking screenshots and do visual regression tests of our pages now so that's a good way for us to ensure that this feature keeps working over time if we actually use it on our site.

Currently there's no highlighting there: 
https://deploy-preview-9183--docusaurus-2.netlify.app/tests/pages/code-block-tests/#code-block-withwithout-the-good-prism-language-case-in-additionallanguages-tests

heysujal committed:
add pHp in additionalLanguages

heysujal committed:
add more test code blocks

heysujal commented on website/_dogfooding/_pages tests/code-block-tests.mdx:
```diff
@@ -27,6 +27,21 @@ See:
 - https://github.com/facebook/docusaurus/pull/3749
 - https://github.com/facebook/docusaurus/pull/6177
 
+## Code block with/without the good prism language case in additionalLanguages[] tests
```
Done

Josh-Cena commented:
I think this should be done within theme config validation, where we map `additionalLanguages` to lowercase in the first place.

slorber commented on packages/docusaurus-theme-classic/src/index.ts:
```diff
@@ -176,7 +176,7 @@ export default function themeClassic(
     configureWebpack() {
       const prismLanguages = additionalLanguages
-        .map((lang) => `prism-${lang}`)
+        .map((lang) => `prism-${lang.toLowerCase()}`)
```
This could be better handled in the config validation, you can use Joi to normalize/lowercase the provided user input values so that everything else in the system see lowercase values (including `prism-include-languages.ts`, it shouldn't be needed to lowercase there too anymore)

slorber commented on website/_dogfooding/_pages tests/code-block-tests.mdx:
````diff
@@ -27,6 +27,41 @@ See:
+?>
+```
+
+```PHP
````
test added, but unfortunately highlighting doesn't work yet

We probably need to lowercase the meta string language in the code block theme component too

slorber reviewed (changes requested):
Thanks, almost good but only 1 of the 3 examples there is highlighted, while I expect the 3 examples to be ;)

https://deploy-preview-9183--docusaurus-2.netlify.app/tests/pages/code-block-tests/#code-block-prism-language-tests

![Uploading CleanShot 2023-07-28 at 19.34.01@2x.png…]()

slorber reviewed (changes requested):
Thanks, almost good but only 1 of the 3 examples there is highlighted, while I expect the 3 examples to be ;)

https://deploy-preview-9183--docusaurus-2.netlify.app/tests/pages/code-block-tests/#code-block-prism-language-tests

<img width="886" alt="CleanShot 2023-07-28 at 19 34 01@2x" src="https://github.com/facebook/docusaurus/assets/749374/d3a3a7a9-9521-47ca-a2df-cc6de0bebe2f">

heysujal commented:
> 

yes we should support the other two as well.

heysujal committed:
Merge branch 'facebook:main' into case-agnostic-highlight

heysujal commented:
@slorber after two weeks I tried working on this PR again. now, I noticed that my code changes are not being picked up as I tried to redo the changes of PR that I have made in the local setup. For example I am using a bad case in additionalLanguages array for php like `'pHP'` and I removed all the toLowerCase() that I added in this PR locally. I expected the code to break. but it still renders without any problem and gives the same output as this.

![image](https://github.com/facebook/docusaurus/assets/55016909/dd8c8251-ca0f-4cb3-88d1-5ad5268afa6e)

After reverting the changes I think it should have broken but it didn't. The changes not being picked up is the reason why I cannot work further on the  meta string thing you mentioned (even though I guessed where to make changes)

I run` yarn workspace website start` whenever I make changes to recompile and build it again.

slorber committed:
Add php title to code blocks, easier to debug

slorber committed:
Revert changes

slorber committed:
Normalize theme prism languages to lowercase

slorber committed:
empty

slorber commented:
Hey, not sure why but maybe:
- you had a local caching issue (run `yarn clear` to be sure)
- you are on a case-insensitive FS (macos, windows) and can't emulate the problem correctly

Anyway, I was able to make it work and the deploy preview is fine, building on Linux, so thanks, that's fine 👍

heysujal commented:
It feels a bit odd to me, I mean you did all the work yet it looks like I have made the changes by creating a PR.

slorber commented:
I just polished a bit what you started so that we can merge 😄 I do that often and it's faster to move on than back and forth PR discussions

heysujal commented:
> I just polished a bit what you started so that we can merge 😄 I do that often and it's faster to move on than back and forth PR discussions

Yes I get it.
